dot/parachain/backing/can_second.go

```diff
@@ -0,0 +1,179 @@
+	hypotheticalCandidate parachaintypes.HypotheticalCandidate,
+	backedInPathOnly bool, //nolint:unparam
+) (map[common.Hash][]uint, error) {
+	var (
+		candidateParaID      parachaintypes.ParaID
+		candidateRelayParent common.Hash
+		candidateHash        parachaintypes.CandidateHash
+		membership           = make(map[common.Hash][]uint)
+	)
+
+	switch v := hypotheticalCandidate.(type) {
+	case parachaintypes.HypotheticalCandidateIncomplete:
+		candidateParaID = v.CandidateParaID
+		candidateRelayParent = v.RelayParent
+		candidateHash = v.CandidateHash
+	case parachaintypes.HypotheticalCandidateComplete:
+		candidateParaID = parachaintypes.ParaID(v.CommittedCandidateReceipt.Descriptor.ParaID)
+		candidateRelayParent = v.CommittedCandidateReceipt.Descriptor.RelayParent
+		candidateHash = v.CandidateHash
+	}
+
+	for head, leafState := range cb.perLeaf {
+		if leafState.prospectiveParachainsMode.IsEnabled {
+
+			// check that the candidate relay parent is allowed for parachain, skip the leaf otherwise.
+			allowedParentsForPara := cb.implicitView.knownAllowedRelayParentsUnder(head, candidateParaID)
+			if !slices.Contains(allowedParentsForPara, candidateRelayParent) {
+				continue
+			}
+
+			responseCh := make(chan parachaintypes.HypotheticalFrontierResponses)
+			cb.SubSystemToOverseer <- parachaintypes.ProspectiveParachainsMessageGetHypotheticalFrontier{
+				HypotheticalFrontierRequest: parachaintypes.HypotheticalFrontierRequest{
+					Candidates:              []parachaintypes.HypotheticalCandidate{hypotheticalCandidate},
+					FragmentTreeRelayParent: &head,
+					BackedInPathOnly:        backedInPathOnly,
+				},
+				Ch: responseCh,
+			}
+
+			res, ok := <-responseCh
+			if ok {
+				var depths []uint
+				// collect all depths from all fragment trees
+				for _, val := range res {
+					for _, membership := range val.Memberships {
+						depths = append(depths, membership.Depths...)
+					}
+				}
+
+				if isSeconded, atDepth := checkDepthsAgainstLeaftState(depths, leafState, candidateParaID); isSeconded {
+					return nil, fmt.Errorf(
+						"%w; candidate hash: %s; relay parent: %s; parachain id: %v; depth: %d",
+						errDepthOccupied,
+						candidateHash.Value.String(),
+						candidateRelayParent.String(),
+						candidateParaID,
+						*atDepth,
+					)
+				}
+
+				membership[head] = depths
+
+			} else {
+				logger.Error("prospective parachains message get hypothetical frontier's response channel is closed")
+			}
+		} else if head == candidateRelayParent {
+			if isSeconded := isSecondedAtDepth(0, leafState, candidateParaID); isSeconded {
+				return nil, fmt.Errorf(
+					"%w; candidate hash: %s; relay parent: %s; parachain id: %v",
+					errLeafOccupied,
+					candidateHash,
+					candidateRelayParent.String(),
+					candidateParaID,
+				)
+			}
+			membership[head] = []uint{0}
+		}
+	}
+
+	if len(membership) == 0 {
+		return nil, fmt.Errorf("%w: %s", errNoActiveLeaves, candidateRelayParent.String())
+	}
+
```

Comment:
Have you addressed this logic https://github.com/paritytech/polkadot-sdk/blob/e2caa813bfd2973e7a7a74aa5292433397dbc787/polkadot/node/core/backing/src/lib.rs#L1193-L1213 in `checkDepthsAgainstLeaftState`?